pypeit/coadd3d.py

```diff
@@ -0,0 +1,1853 @@
+        * A common spaxel scale is determined, and the astrometric correction is derived
+        * An RA and Dec image is created for each pixel.
+        * Based on atmospheric conditions, a differential atmospheric refraction correction is applied.
+        * Extinction correction
+        * Flux calibration (optional - this calibration is only applied if a standard star cube is supplied)
+
+        If the input frames will not be combined (combine=False) if they won't be aligned (align=False), then
+        each individual spec2d file is converted into a spec3d file (i.e. a PypeIt DataCube object). These fits
+        files can be loaded/viewed in other software packages to display or combine multiple datacubes into a
+        single datacube. However, note that different software packages use combination algorithms that may not
+        conserve flux, or may produce covariance between adjacent voxels.
+
+        If the user wishes to either spatially align multiple exposures (align=True) or combine multiple
+        exposures (combine=True), then the next set of operations include:
+
+        * Generate white light images of each individual cube (according to a user-specified wavelength range)
+        * Align multiple frames if align=True (either manually by user input, or automatically by cross-correlation)
+        * Create the output WCS, and apply the flux calibration to the data
+        * Generate individual datacubes (combine=False) or one master datacube containing all exposures (combine=True).
+          Note, there are several algorithms used to combine multiple frames. Refer to the subpixellate() routine for
+          more details about the combination options.
+        """
+        # No need to continue if we are not combining nor aligning frames
+        if not self.combine and not self.align:
+            return
+
+        # If the user is aligning or combining, the spatial scale of the output cubes needs to be consistent.
+        # Set the spatial and spectral scales of the output datacube
+        self.set_voxel_sampling()
+
+        # Align the frames
+        if self.align:
+            self.run_align()
+
+        # Compute the relative weights on the spectra
+        self.all_wghts = self.compute_weights()
+
+        # Generate the WCS, and the voxel edges
+        cube_wcs, vox_edges, _ = self.create_wcs(self.all_ra, self.all_dec, self.all_wave, self._dspat, self._dwv)
+
+        sensfunc = None
+        if self.flux_spline is not None:
+            # Get wavelength of each pixel
+            numwav = vox_edges[2].size - 1
+            wcs_scale = (1.0 * cube_wcs.spectral.wcs.cunit[0]).to(units.Angstrom).value  # Ensures the WCS is in Angstroms
+            senswave = wcs_scale * cube_wcs.spectral.wcs_pix2world(np.arange(numwav), 0)[0]
+            sensfunc = self.flux_spline(senswave)
+
+        # Generate a datacube
+        outfile = datacube.get_output_filename("", self.cubepar['output_filename'], True, -1)
```

Comment:
My advice is to create the this outfile name int he __init__ method or before coadd is called. And then have self.coadd(outfile) be the calling sequence, to make it clear that this coadds and writes to disk.